lib/experimental/modules/class-gutenberg-modules.php

```diff
@@ -0,0 +1,186 @@
+<?php
+/**
+ * Modules API: Gutenberg_Modules class.
+ *
+ * Native support for ES Modules and Import Maps.
+ *
+ * @package Gutenberg
+ * @subpackage Modules
+ */
+
+/**
+ * Gutenberg_Modules class
+ */
+class Gutenberg_Modules {
+	/**
+	 * An array of registered modules, keyed by module identifier.
+	 *
+	 * @var array
+	 */
+	private static $registered = array();
+
+	/**
+	 * An array of queued modules.
+	 *
+	 * @var string[]
+	 */
+	private static $enqueued = array();
+
+	/**
+	 * Registers the module if no module with that module identifier already
+	 * exists.
+	 *
+	 * @param string           $module_identifier The identifier of the module. Should be unique. It will be used in the final import map.
+	 * @param string           $src               Full URL of the module, or path of the script relative to the WordPress root directory.
+	 * @param array            $dependencies      Optional. An array of module identifiers of the static and dynamic dependencies of this module. It can be an indexed array, in which case all the dependencies are static, or it can be an associative array, in which case it has to contain the keys `static` and `dynamic`.
```

Comment:
I wonder if instead of having subkey here  to separate these two into two top level arrays (though harder to name them)

Reply:
I guess my question also relates to the fact that if we ever want to move from "server graph" to "no server graph" at some point, the API would stay the same (no breaking change), we'd just change how we interpret it.

Reply:
So, you mean something like this, right?

```
wp_register_module(
  - string $module_identifier,
  - string|false $src,
  - string[] $static_deps = array(),
  - string[] $dynamic_deps = array(),
  - string|bool|null $ver = false
)
```

Sure, we could do that. My point about keeping a single array with optional subkeys was:

- Dynamic dependencies are less common than static ones, so you wouldn't need to use subkeys that often.
- It's closer to `wp_register_script`, which is something people are familiar with.
- If someone adds a dynamic dependency as a static one (forgets using subkeys), it's not optimal, but it's not broken. Also, they will see a warning in the devtools console.

Reply:
what about the point about making both experiments compatible in terms of APIs?

Reply:
What do you mean? So we can hypothetically move from option 3 (required dependencies) to option 2 (optional dependencies) in the future?

Reply:
yes, maybe it's not that important thought and we should just make the right decision :)

Reply:
By the way, there's an alternative API that I don't dislike because it opens the door for other dependency properties in the future:

- Single dependencies array:
  ```php
  wp_register_module( 'my-module', '/src', array( 'static-dep' ) );
  ```
- But dependencies can be strings or arrays if they have properties:
  ```php
  wp_register_module(
    'my-module',
    '/src',
    array(
      'static-dep',
      array( 'id' => 'dynamic-dep', 'type' => 'dynamic' ),
    )
  );
  ```
- Other properties could be added if necessary in the future, for example, if we'd ever want to support `scopes`:
  ```php
  wp_register_module(
    'my-module',
    '/src',
    array(
      array( 'id' => 'static-dep', 'version' => '^3.0.0' ),
    )
  );
  ```
- This one also prioritizes the simplicity of the most common case, which is static dependencies.

Reply:
We are not using any dynamic dependency yet, so I'll make this change in a subsequent PR.